### PR TITLE
Add validation of generation and metageneration to upload functions

### LINKED ISSUE #16: Storage: upload_from_string() with ifGenerationMatch=0
The GCS HTTP protocol -- but not the Python API -- has the ability to set `ifGenerationMatch` when [creating a storage object](https://cloud.google.com/storage/docs/json_api/v1/objects/insert#parameters):

> Makes the operation conditional on whether the object's current generation matches the given value. Setting to 0 makes the operation succeed only if there is a live version of the object.

**Why it's useful:** With this feature, the client could create a directory placeholder entry (a 0-byte object with a name ending in `'/'`) very efficiently like this:

    blob = bucket.blob('path/to/my/subdirectory/')
    blob.upload_from_string(b'', if_generation_match=0)

That one round trip creates the directory placeholder entry if it doesn't already exist. The alternatives are to first make a round trip to check if the entry exists or else to let the bucket accumulate identical placeholder entries (esp. for top level directories) by blindly creating them. [<strike>Or does GCS check if an uploaded object matches the current generation and optimize that case?</strike> -- Nope.]

**Why that matters:** Directory placeholders speed up [gcsfuse](https://github.com/GoogleCloudPlatform/gcsfuse) by an order of magnitude. Without the placeholders, you have to use `gcsfuse` in `--implicit-dirs` mode, and such a mount is frustratingly slow for interactive work. E.g. it takes several seconds just to list a tiny directory containing 2 files. With the placeholders, you can run `gcsfuse` without `--implicit-dirs`, and that mount lists directories in a tenth of a second or two.

**Proposal:** I could create a Pull Request adding this feature if you like, with either the specific `if_generation_match` query parameter or a way to pass in additional query parameters.

Another alternative is recommend that callers do something like subclass `Blob` to override `_add_query_parameters()` to add the `if_generation_match=0` name-value pair. That's ugly and fragile.

Is there a way to do this that I'm missing? Are there better alternatives?

### PR DESCRIPTION
**Pull request by HollayHorvath**
- [x] An issue was already open for this problem, linked below
- [x] Ensure the tests and linter pass
- [ ] Code coverage does not decrease - I'm not quite sure about this
- [x] Appropriate docs were updated - I have updated the comments by adding the new parameters, if there are other places to update please let me know

Fixes #16  🦕

### CODE REVIEW

**Comment by HollayHorvath** [comment]
@googlebot I signed it!

File: google/cloud/storage/blob.py [hunk 1]
```diff
@@ -966,7 +965,17 @@ def _get_upload_arguments(self, content_type):
...
     def _do_multipart_upload(
-        self, client, stream, content_type, size, num_retries, predefined_acl
+        self,
+        client,
+        stream,
+        content_type,
+        size,
+        num_retries,
+        predefined_acl,
+        generation_match,
```
Comment (frankyn) [inline google/cloud/storage/blob.py:975]: Could you update the name with expected prefix of `if_` to follow the API spec?

File: tests/unit/test_blob.py [hunk 2]
```diff
@@ -1785,6 +1897,10 @@ def _upload_from_file_helper(self, side_effect=None, **kwargs):
         content_type = u"font/woff"
         client = mock.sentinel.client
         predefined_acl = kwargs.get("predefined_acl", None)
+        generation_match = kwargs.get("generation_match", None)
```
Comment (frankyn) [inline tests/unit/test_blob.py:1900]: One more request, could you add system tests that exercise these new parameters?

**Comment by frankyn** [comment]
Thanks for your support @HollayHorvath, you helped get this ball moving and will have it out in the library soon. 

@IlyaFaer is taking the work you've done and making necessary changes to get into the library. I'm closing out this PR and continuing the remaining discussions in #111.

**Pull request closed without merging**